climada/hazard/trop_cyclone.py

```diff
@@ -387,38 +387,69 @@ def _tc_from_track(self, track, centroids, coastal_idx, model='H08',
-                new_val = getattr(haz_cc, chg['variable'])
-                # 1d-masks like `select` are inefficient for indexing sparse matrices since
-                # they are broadcasted densely in the second dimension
-                if isinstance(new_val, sparse.csr_matrix):
-                    new_val = sparse.diags(np.where(select, change, 1)).dot(new_val)
-                else:
-                    new_val[select] *= change
-                setattr(haz_cc, chg['variable'], new_val)
-        return haz_cc
+        tc_cc = copy.deepcopy(self)
+        tc_cc_basin = np.array(tc_cc.basin)
+        
+        #criterion per basin
+        for basin in np.unique(tc_cc_basin):
+        
+            basin_chg = [chg for chg in chg_int_freq if chg['basin'] == basin]
+            bas_sel = (tc_cc_basin == basin)
+
+            #Apply intensity change
+            inten_chg = [chg
+                         for chg in basin_chg
+                         if chg['variable'] =='intensity'
+                         ]
+            for chg in inten_chg:
+                sel_cat_chg = np.isin(tc_cc.category, chg['category']) & bas_sel
+                inten_scaling = 1 + (chg['change'] - 1) * scaling_rcp_year
+                tc_cc.intensity = sparse.diags(
+                    np.where(sel_cat_chg, inten_scaling, 1)
+                    ).dot(tc_cc.intensity)
+            
+            #Apply frequency change
+            freq_chg = [chg 
+                        for chg in basin_chg
+                        if chg['variable']=='frequency'
+                        ]
+            freq_chg.sort(reverse = False, key = lambda x: len(x['category']))
+            
+
+            #Iteratively scale frequencies for each category such that 
+            #cumulative frequencies are scaled according to Knuston criterion.
+
+            cat_larger_list = []
+            for chg in freq_chg:
+                cat_chg_list = [cat 
+                                for cat in chg['category']
+                                if cat not in cat_larger_list
+                                ]
+                sel_cat_chg = np.isin(tc_cc.category, cat_chg_list) & bas_sel
+                if sel_cat_chg.any():
+                    freq_scaling = 1 + (chg['change'] - 1) * scaling_rcp_year
+                    sel_cat_all = (np.isin(tc_cc.category, chg['category']) 
+                                   & bas_sel)
+                    sel_cat_larger = (np.isin(tc_cc.category, cat_larger_list)
+                                      & bas_sel)
+                    freq_scaling_cor = (
+                        (np.sum(self.frequency[sel_cat_all]) * freq_scaling
+                         - np.sum(tc_cc.frequency[sel_cat_larger]))
+                        / np.sum(self.frequency[sel_cat_chg])
+                    )
+                    tc_cc.frequency[sel_cat_chg] *= freq_scaling_cor
+                cat_larger_list += cat_chg_list
+                
+        return tc_cc
+    
```

Comment:
Remove trailing white space (https://www.python.org/dev/peps/pep-0008/#other-recommendations)
```suggestion

        return tc_cc

```